applications/FluidDynamicsApplication/custom_elements/surface_smoothing_element.cpp

```diff
@@ -0,0 +1,670 @@
+ */
+void SurfaceSmoothingElement::CalculateLocalSystem(
+    MatrixType& rLeftHandSideMatrix,
+    VectorType& rRightHandSideVector,
+    ProcessInfo& rCurrentProcessInfo)
+{
+    KRATOS_TRY
+
+    const double dt = rCurrentProcessInfo.GetValue(DELTA_TIME);
+    const double he = ElementSizeCalculator<3,4>::AverageElementSize(GetGeometry());
+    const double epsilon = 1.0e2*dt*he*he;
+
+    BoundedMatrix<double,num_nodes,num_dim> DN_DX;  // Gradients matrix
+    array_1d<double,num_nodes> N; // dimension = number of nodes . Position of the gauss point
+
+    array_1d<double,num_nodes> PHIdof; //dimension = number of DOFs, needed since we are using a residualbased approach
+    array_1d<double,num_nodes> PHIold; //dimension = number of DOFs
+    array_1d<VectorType,num_nodes> GradPHIold;
+
+    BoundedMatrix<double,num_nodes,num_nodes> tempM;
+    tempM = ZeroMatrix(num_nodes,num_nodes);
+    BoundedMatrix<double,num_nodes,num_nodes> tempA;
+    tempA = ZeroMatrix(num_nodes,num_nodes);
+
+    array_1d<double,num_nodes> tempBCRHS = ZeroVector(num_nodes);
+
+    if(rLeftHandSideMatrix.size1() != num_nodes) // be carefull that num_dof = num_nodes
+        rLeftHandSideMatrix.resize(num_nodes,num_nodes,false); //resizing the system in case it does not have the right size
+
+    if(rRightHandSideVector.size() != num_nodes)
+        rRightHandSideVector.resize(num_nodes,false);
+
+    // Getting data for the given geometry
+    double volume;
+    GeometryUtils::CalculateGeometryData(GetGeometry(), DN_DX, N, volume); //asking for gradients and other info
+
+    for(unsigned int i = 0; i<num_nodes; i++){
+        PHIold[i] = GetGeometry()[i].GetValue(DISTANCE);
+        PHIdof[i] = GetGeometry()[i].FastGetSolutionStepValue(DISTANCE);
+        GradPHIold[i] = GetGeometry()[i].FastGetSolutionStepValue(DISTANCE_GRADIENT);
+    }
+
+    for(unsigned int i = 0; i<num_nodes; i++){
+        for(unsigned int j = 0; j<num_nodes; j++){
+            tempM(i,j) = volume*N[i]*N[j];
+
+            for (unsigned int k = 0; k<num_dim; k++){
+                tempA(i,j) += volume*epsilon*DN_DX(i,k)*DN_DX(j,k);
+            }
+        }
+    }
+
+    // Implementing the boundary condition on distance gradient: can be implemented
+    // by a custom condition for a more general case
+
+    const auto& neighbour_elems = this->GetValue(NEIGHBOUR_ELEMENTS);
+
+    for (unsigned int i_ne = 0; i_ne < num_faces; i_ne++){
+        if (neighbour_elems[ i_ne ].Id() == this->Id() ){
+            auto outer_face = Triangle3D3< GeometryType::PointType >(
+                                    this->pGetGeometry()->pGetPoint(mNode0ID[i_ne]),
+                                    this->pGetGeometry()->pGetPoint(mNode1ID[i_ne]),
+                                    this->pGetGeometry()->pGetPoint(mNode2ID[i_ne]));
+
+            unsigned int contact_node = 0;
+            for (unsigned int i=0; i < num_face_nodes; ++i){
+                if ( outer_face[i].GetValue(IS_STRUCTURE) == 1.0 ){
```

Comment:
```suggestion
                if (outer_face[i].Is(STRUCTURE)){
```

this is what you should do, i.e. this is the interface of `Kratos::Flags`

I will explain more later

in the meantime  I recommend checking how the `STRUCTURE` and the `SLIP` flags are used in FluidDynamics

Reply:
@philbucher thanks. Actually I am familiar with using kratos flags. My point was to choose a proper one describing my case. 

- SLIP is reserved for slip boundary, which should be used for my test-case (later).
- STRUCTURE might create some complexities in some test-cases (not sure but maybe in the future someone wants to use two-fluids-solver in FSI cases?)
- CONTACT seems fine for me but I do not know where it has been used before to prevent any unintentional conflict.

Just describing the solver that I am working on to make it more clear:
It is an extension to the two fluids Navier Stokes solver considering surface tension and liquid-gas-solid contact forces. I need to mark nodes that coincide with the solid surface (on which droplets are spreading). At the moment I use IS_STRUCTURE variable to consistently mark solid surface in all subroutines. If I decide to substitute IS_STRUCTURE, I should use a unique flag/variable to prevent any conflict at a later stage.

Reply:
@philbucher thank you for raising the issue.
Done (CONTACT flag is used).

Reply:
cool
@rubenzorrilla are you ok with using `CONTACT` here?

Reply:
Sure, I think it makes sense.